auth/client/src/main/java/org/wildfly/security/auth/client/EncryptedExpressionResolver.java

```diff
@@ -0,0 +1,195 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2024 Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.wildfly.security.auth.client;
+
+import org.wildfly.security.credential.SecretKeyCredential;
+import org.wildfly.security.credential.store.CredentialStore;
+import org.wildfly.security.credential.store.CredentialStoreException;
+
+import javax.crypto.SecretKey;
+import java.security.GeneralSecurityException;
+import java.util.Collections;
+import java.util.Map;
+
+import static org.wildfly.common.Assert.checkNotNullParam;
+import static org.wildfly.security.auth.client._private.ElytronMessages.xmlLog;
+import static org.wildfly.security.encryption.CipherUtil.decrypt;
+import static org.wildfly.security.encryption.CipherUtil.encrypt;
+
+
+/**
+ * A class used to resolve encrypted expressions using secret key within credential stores.
+ * Contains a collection of multiple resolver configurations.
+ *
+ * @author <a href="mailto:prpaul@redhat.com">Prarthona Paul</a>
+ */
+
+public class EncryptedExpressionResolver {
+    private static final String CREDENTIAL_STORE_API_CAPABILITY = "org.wildfly.security.credential-store-api";
+    private volatile boolean initialised = false;
+    private final ThreadLocal<String> initialisingFor = new ThreadLocal<>();
+    private volatile ExpressionResolutionException firstFailure = null;
+
+    private volatile String prefix;
+    private volatile String completePrefix;
+    private volatile String defaultResolver;
+    private volatile Map<String, ResolverConfiguration> resolverConfigurations;
+
+    public EncryptedExpressionResolver() {
+//        this.configurator = configurator;
+    }
+
+    public String resolveExpression(String expression, EncryptedExpressionConfig config) {
+        checkNotNullParam("expression", expression);
+        checkNotNullParam("encrypted expression configuration", config);
+        return resolveExpressionInternal(expression, config);
+    }
+
+    public Map<String, ResolverConfiguration> getResolverConfiguration() {
+        return resolverConfigurations;
+    }
+
+    private String resolveExpressionInternal(String fullExpression, EncryptedExpressionConfig config) {
+        assert config != null;
+
+        if (fullExpression.length() > 3) {
```

Comment:
@PrarthonaPaul most code in this class seems the same as in the wildfly-core repo. Can you extract the parts of the code used by both to some Utils class and make both wildfly-elytron-client and wildfly-core use the methods from Utils, so we avoid code duplication?

Reply:
I am not sure if we could do this (at least for some of the bigger ones, like createExpression or resolveExpressionInternal or resolveExpression since on the core side, they are server dependent). 
But some of the smaller functions, like setResolverConfigurations or setDefaultResolver or setPrefix, we can add those to the utility.

Reply:
@PrarthonaPaul The methods that can be extracted to common Utils should be extracted, so that if a bug is found in the future, we can fix it on a single place. You can also refactor the methods on the server side to be able to introduce common Utils methods